doc/requirements.txt

```diff
@@ -0,0 +1,2 @@
+sphinx==3.4.3
+sphinx-rtd-theme==0.5.1
```

Comment:
These are just for RTD. How would you like to declare documentation dependencies? In `beartype.meta` as a separate category like `LIBS_DEVELOPMENT`?

Reply:
**Right.** This is the standard approach and for good reason. But I'm a dependency wierdo, as you've probably already observed from `beartype.meta`.

The `doc/requirements.txt` approach is *absolutely* fine. We can run with this as is. My personal preference, though, would be to define a top-level `.readthedocs.yml` configuration file resembling:

```
version: 2

python:
  version: 3.7
  install:
    - method: pip
      path: .
      extra_requirements:
        - doc
  system_packages: true
```

RTD would then install `beartype` as:

```
$ pip install .[docs]
```

This is kinda nice, because it enables end users to do the same. Maybe someone wants that? Who knows! But I'd like to give that to them, if we can.

To facilitate this, I'll need to do a bit of additional legwork on my end. But that's fine! I have legs, so legwork is feasible. This includes:

* Defining a new `beartype.meta.LIBS_DOCTIME_MANDATORY` tuple resembling:

```
LIBS_DOCTIME_MANDATORY = {
    'sphinx >=3.4.3',
    'sphinx-rtd-theme >=0.5.1',
}
```

* Defining a new `'extras_require'` key-value pair in `setup` resembling:

```
        'doc': meta.LIBS_DOCTIME_MANDATORY,
```

Don't feel to obliged to do *any* of that. That's entirely my obligation.

Here's where you come in! My main question is the pinning of dependencies in `doc/requirements.txt`. I usually like to avoid pinning in open-source projects to:

* Improve forward compatibility with a wider range of versions.
* Enable end users to install dependencies with their preferred package manager (e.g., `apt`), which often fail to satisfy specific versions.
* Simplify my life by notifying me with [failing badges](https://docs.readthedocs.io/en/stable/badges.html) and automated e-mails when dependency bumps break our documentation build.

Does all of that make sense or am I inviting a horrifying reenactment of Pompeii by failing to pin versions?

Reply:
Yep! I mostly agree. In my mind pinning exact versions for RTD is more like a pin for deployment, rather than the library, for which I wholeheartedly second more loose requirements. But perhaps `sphinx ~= 3.4` and such could be appropriate instead of simply `>=`? Your call! And having users do `pip install -e .[docs,test]` vs `-e .[dev]` could be an argument to be had as well. [I've used](https://github.com/felix-hilden/tekore/blob/master/setup.py#L36) a `dev` collection to gather all the different extras into one, which seems to be common practice. But two isn't that bad. Three gets a bit annoying.

Reply:
> In my mind pinning exact versions for RTD is more like a pin for deployment...

This makes rationality. On the one sweaty hand, duplicating requirements across `doc/requirements.txt` and `beartype/meta.py` [violates DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself); on the other sweaty hand, *not* pinning RTD requirements invites build failures for no particularly good reason.

*Ugh, choices.* O.K., I'm sold. Let's duplicate requirements as you say, because breaking documentation builds for no particularly good reason is word. Also, your word is law.

> And having users do `pip install -e .[docs,test]` vs `-e .[dev]` could be an argument to be had as well.

You're... right, actually. If I was a contributor, I'd personally be annoyed to the point of rabidly gnashing teeth at even two different extras. Sadly, the `test` extra is required by our top-level `tox` configuration:

```
extras =
    # Install all optional runtime dependencies to maximize test coverage.
    all

    # Install all mandatory test-specific dependencies. This is the official
    # solution supported by "tox" developers for eliminating redundancy between
    # testing dependencies listed within this file and the top-level "setup.py"
    # script. While non-intuitive, we have little recourse. See also:
    #     https://stackoverflow.com/questions/29870629/pip-install-test-dependencies-for-tox-from-setup-py
    #     https://stackoverflow.com/questions/39922650/tox-tests-use-setup-py-extra-require-as-tox-deps-source
    #     https://github.com/tox-dev/tox/issues/13#issuecomment-247788280
    # Note that this also requires ".[test]" to be listed as a dependency.
    test
```

*Ugh.* But that's fine. We can additionally define a new `dev` extra defined as the tuple union of `test` and `doc`, enabling contributors to trivially `pip install -e .[dev]` without the fist-clenching hassle of extra extras. For example, in `setup.py`:

        'dev': meta.LIBS_TESTTIME_MANDATORY + meta.LIBS_DOCTIME_MANDATORY,

Behold, the horrifying black magic of `beartype.meta`! :mage_man: